constraint-solver/src/quadratic_symbolic_expression.rs

```diff
@@ -603,39 +597,24 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display> QuadraticSymbolicExp
     fn transfer_constraints(
         &self,
         range_constraints: &impl RangeConstraintProvider<T, V>,
-    ) -> Option<Effect<T, V>> {
-        // We are looking for X = a * Y + b * Z + ... or -X = a * Y + b * Z + ...
-        // where X is least constrained.
-
+    ) -> Vec<Effect<T, V>> {
+        // Solve for each of the variables in the linear component and
+        // compute the range constraints.
         assert!(!self.is_quadratic());
-
-        let (solve_for, solve_for_coefficient) = self
-            .linear
-            .iter()
-            .filter(|(_var, coeff)| coeff.is_known_one() || coeff.is_known_minus_one())
-            .max_by_key(|(var, _c)| {
-                // Sort so that we get the least constrained variable.
-                range_constraints.get(var).range_width()
-            })?;
-
-        // This only works if the coefficients are all known.
-        let summands = self
-            .linear
+        self.linear
             .iter()
-            .filter(|(var, _)| *var != solve_for)
-            .map(|(var, coeff)| {
-                let coeff = coeff.try_to_number()?;
-                Some(range_constraints.get(var).multiple(coeff))
+            .filter(|(_, coeff)| coeff.is_known_nonzero())
+            .map(|(var, _)| {
+                (
+                    var,
+                    self.try_solve_for(var)
+                        .unwrap()
+                        .range_constraint(range_constraints),
+                )
             })
```

Comment:
```suggestion
            .filter_map(|(var, _)| {
                self.try_solve_for(var)
                    .map(|expr| (var, expr.range_constraint(range_constraints)))
            })
```
I think this should be equivalent?

Reply:
Yes, it saves a clone. But let me actually do the change in `try_solve_for`.